larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc

```diff
@@ -0,0 +1,125 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc
+ *
+ *  @brief  Implementation file for the simple PCA-based clustering tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArPcaHelper.h"
+#include "larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.h"
+#include "larpandoracontent/LArHelpers/LArPointingClusterHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+SimplePCAThreeDClusteringTool::SimplePCAThreeDClusteringTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool SimplePCAThreeDClusteringTool::Run(const Algorithm *const /*pAlgorithm*/, std::vector<pandora::CaloHitList*> &newCaloHitListsVector)
+{
+    if (newCaloHitListsVector.size() != 1)
+        return false;
+
+    CaloHitList *pInitialCaloHitList = newCaloHitListsVector.at(0);
+    newCaloHitListsVector.clear();
+
+    // Begin with a PCA
+    CartesianVector centroid(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecs;
+    LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(*pInitialCaloHitList, centroid, eigenValues, eigenVecs);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirection(eigenVecs.at(0).GetZ() > 0.f ? eigenVecs.at(0) : eigenVecs.at(0) * -1.f);
+
+    // Place intercept at hit with minimum projection
+    float minProjection(std::numeric_limits<float>::max());
+    for (const CaloHit *const pCaloHit3D : *pInitialCaloHitList)
+        minProjection = std::min(minProjection, axisDirection.GetDotProduct(pCaloHit3D->GetPositionVector() - centroid));
+
+    const CartesianVector axisIntercept(centroid + (axisDirection * minProjection));
+
+    // Now define ortho directions
+    const CartesianVector seedDirection((axisDirection.GetX() < std::min(axisDirection.GetY(), axisDirection.GetZ())) ? CartesianVector(1.f, 0.f, 0.f) :
+        (axisDirection.GetY() < std::min(axisDirection.GetX(), axisDirection.GetZ())) ? CartesianVector(0.f, 1.f, 0.f) : CartesianVector(0.f, 0.f, 1.f));
+    const CartesianVector orthoDirection1(seedDirection.GetCrossProduct(axisDirection).GetUnitVector());
```

Comment:
Maybe I'm misunderstanding the intent here, but can you not get the ortho direction from the second eigenvector directly?

Reply:
Oh yeah absolutely!